includes/class-wc-stripe-webhook-handler.php

```diff
@@ -106,36 +106,49 @@ public function validate_request( $request_headers, $request_body ) {
 			return WC_Stripe_Webhook_State::VALIDATION_FAILED_EMPTY_BODY;
 		}
 
-		if ( ! empty( $request_headers['USER-AGENT'] ) && ! preg_match( '/Stripe/', $request_headers['USER-AGENT'] ) ) {
-			return WC_Stripe_Webhook_State::VALIDATION_FAILED_USER_AGENT_INVALID;
+		if ( empty( $this->secret ) ) {
+			return $this->validate_request_user_agent( $request_headers );
 		}
 
-		if ( ! empty( $this->secret ) ) {
-			// Check for a valid signature.
-			$signature_format = '/^t=(?P<timestamp>\d+)(?P<signatures>(,v\d+=[a-z0-9]+){1,2})$/';
-			if ( empty( $request_headers['STRIPE-SIGNATURE'] ) || ! preg_match( $signature_format, $request_headers['STRIPE-SIGNATURE'], $matches ) ) {
-				return WC_Stripe_Webhook_State::VALIDATION_FAILED_SIGNATURE_INVALID;
-			}
+		// Check for a valid signature.
+		$signature_format = '/^t=(?P<timestamp>\d+)(?P<signatures>(,v\d+=[a-z0-9]+){1,2})$/';
+		if ( empty( $request_headers['STRIPE-SIGNATURE'] ) || ! preg_match( $signature_format, $request_headers['STRIPE-SIGNATURE'], $matches ) ) {
+			return WC_Stripe_Webhook_State::VALIDATION_FAILED_SIGNATURE_INVALID;
+		}
 
-			// Verify the timestamp.
-			$timestamp = intval( $matches['timestamp'] );
-			if ( abs( $timestamp - time() ) > 5 * MINUTE_IN_SECONDS ) {
-				return WC_Stripe_Webhook_State::VALIDATION_FAILED_TIMESTAMP_MISMATCH;
-			}
+		// Verify the timestamp.
+		$timestamp = intval( $matches['timestamp'] );
+		if ( abs( $timestamp - time() ) > 5 * MINUTE_IN_SECONDS ) {
+			return WC_Stripe_Webhook_State::VALIDATION_FAILED_TIMESTAMP_MISMATCH;
+		}
 
-			// Generate the expected signature.
-			$signed_payload     = $timestamp . '.' . $request_body;
-			$expected_signature = hash_hmac( 'sha256', $signed_payload, $this->secret );
+		// Generate the expected signature.
+		$signed_payload     = $timestamp . '.' . $request_body;
+		$expected_signature = hash_hmac( 'sha256', $signed_payload, $this->secret );
 
-			// Check if the expected signature is present.
-			if ( ! preg_match( '/,v\d+=' . preg_quote( $expected_signature, '/' ) . '/', $matches['signatures'] ) ) {
-				return WC_Stripe_Webhook_State::VALIDATION_FAILED_SIGNATURE_MISMATCH;
-			}
+		// Check if the expected signature is present.
+		if ( ! preg_match( '/,v\d+=' . preg_quote( $expected_signature, '/' ) . '/', $matches['signatures'] ) ) {
+			return WC_Stripe_Webhook_State::VALIDATION_FAILED_SIGNATURE_MISMATCH;
 		}
 
 		return WC_Stripe_Webhook_State::VALIDATION_SUCCEEDED;
 	}
 
+	/**
+	 * Verify User Agent of the incoming webhook notification. Used as fallback for the cases when webhook secret is missing.
+	 *
+	 * @since 5.0.0
+	 * @version 5.0.0
+	 * @param array $request_headers The request headers from Stripe.
+	 * @return string The validation result (e.g. self::VALIDATION_SUCCEEDED )
+	 */
+	private function validate_request_user_agent( $request_headers ) {
+		$ua_is_valid = empty( $request_headers['USER-AGENT'] ) || preg_match( '/Stripe/', $request_headers['USER-AGENT'] );
```

Comment:
Looks like the previous implementation allowed the User-Agent to be empty with
```php
! empty( $request_headers['USER-AGENT'] ) && ! preg_match( '/Stripe/', $request_headers['USER-AGENT'] )
```

But now we shouldn't, otherwise it would be invalid, right?

```php
! empty( $request_headers['USER-AGENT'] ) && preg_match( '/Stripe/', $request_headers['USER-AGENT'] );
```

Reply:
I've left our own check the same. Previous check was for invalid UA, I've invertied it (`! a && ! b === ! (a || b)`) to check if it is valid.

> Looks like the previous implementation allowed the User-Agent to be empty with

So does this implementation.

> But now we shouldn't, otherwise it would be invalid, right?

I don't see the reason why the UA shouldn't be empty. I'm not sure if Stripe always sets UA header.

Reply:
I thought this PR would enforce the UA to be valid if webhook secret is empty.

> I don't see the reason why the UA shouldn't be empty. I'm not sure if Stripe always sets UA header.

🤔 IDK either... I think it's safer to leave it this way.

Reply:
> I thought this PR would enforce the UA to be valid if webhook secret is empty.

It kind of does, but empty UA was considered as valid, hence I haven't touched that part.